core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1197,6 +1213,71 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     KVUtils.open(newStorePath, metadata)
   }
 
+  private def createHybridStore(
+      dm: HistoryServerDiskManager,
+      appId: String,
+      attempt: AttemptInfoWrapper,
+      metadata: AppStatusStoreMetadata): KVStore = {
+
+    val reader = EventLogFileReader(fs, new Path(logDir, attempt.logPath),
+      attempt.lastIndex)
+    val isCompressed = reader.compressionCodec.isDefined
+
+    val memoryUsage = approximateMemoryUsage(reader.totalSize, isCompressed)
+    if (currentInMemoryStoreUsage.get + memoryUsage > maxInMemoryStoreUsage) {
+      throw new IllegalStateException("Not enough in-memory storage to create hybrid store.")
+    }
+    currentInMemoryStoreUsage.addAndGet(memoryUsage)
+    logInfo(s"Attempt creating hybrid store to parse $appId / ${attempt.info.attemptId}. " +
+      s"Requested ${Utils.bytesToString(memoryUsage)} in-memory storage quota.")
+
+    logInfo(s"Leasing disk manager space for app $appId / ${attempt.info.attemptId}...")
+    val lease = dm.lease(reader.totalSize, isCompressed)
+    val isLeaseRolledBack = new java.util.concurrent.atomic.AtomicBoolean(false)
+    var store: HybridStore = null
+    try {
+      store = new HybridStore()
+      val levelDB = KVUtils.open(lease.tmpPath, metadata)
+      store.setLevelDB(levelDB)
+      store.setCachedQuantileKlass(classOf[CachedQuantile])
```

Comment:
Since HybridStore depends on the logic in AppStatusStore (hence specific to SHS), I think it's not required for HybridStore to be placed in common-kvstore module. HybridStore can be moved to core module (this kind of workaround is not needed then) and there's another chance you can simplify the code (+ test code) via implementing to Scala.

Reply:
ok, will do that. I am moving it to org.apache.spark.deploy.history module.